src/lib.rs

```diff
@@ -569,15 +635,64 @@ impl<'a, 'mir, 'tcx> Machine<'a, 'mir, 'tcx> for Evaluator<'tcx> {
     #[inline(always)]
     fn stack_push(
         ecx: &mut InterpretCx<'a, 'mir, 'tcx, Self>,
-    ) -> EvalResult<'tcx, stacked_borrows::CallId> {
-        Ok(ecx.memory().extra.borrow_mut().new_call())
+    ) -> EvalResult<'tcx, FrameData<'tcx>> {
+        Ok(FrameData {
+            call_id: ecx.memory().extra.borrow_mut().new_call(),
+            catch_panic: None,
+            is_box_me_frame: false
+        })
     }
 
     #[inline(always)]
     fn stack_pop(
         ecx: &mut InterpretCx<'a, 'mir, 'tcx, Self>,
-        extra: stacked_borrows::CallId,
-    ) -> EvalResult<'tcx> {
-        Ok(ecx.memory().extra.borrow_mut().end_call(extra))
+        extra: FrameData,
+    ) -> EvalResult<'tcx, StackPopInfo> {
+        if extra.is_box_me_frame {
+            trace!("unwinding: found box_me_frame");
+            ecx.machine.unwinding = true;
+        }
+        if ecx.machine.unwinding {
+            trace!("Popping during unwind!");
+            if let Some(unwind_data) = ecx.frame_mut().extra.catch_panic.take() {
+                // We've just popped the frame that was immediately above
+                // our target frame on the stack.
+                //
+                trace!("unwinding: found target frame: {:?}", ecx.frame().span);
+
+                // 'box_me_up' has finished. 'temp_ptr' now holds
+                // a '*mut (dyn Any + Send)'
+                // We want to split this into its consituient parts -
+                // the data and vtable pointers - and store them back
+                // into the panic handler frame
+                let tmp_ptr = ecx.machine.box_me_up_tmp_ptr.take().unwrap();
```

Comment:
Oh, I did not realize that `box_me_up_tmp_ptr` is used to return things from `box_me_up` to `catch_panic`! That's "magic action at a distance". Is that really needed? How does the passing-on of exception data work in real unwinding implementations?